nemo_curator/classifiers/aegis.py

```diff
@@ -0,0 +1,271 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+
+os.environ["RAPIDS_NO_INITIALIZE"] = "1"
+os.environ["DASK_DATAFRAME__QUERY_PLANNING"] = "False"
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import List, Optional, Union
+
+import cudf
+import torch
+import torch.nn as nn
+from crossfit import op
+from crossfit.backend.torch.hf.model import HFModel
+from peft import PeftModel
+from transformers import AutoConfig, AutoModelForCausalLM, AutoTokenizer
+
+from nemo_curator.classifiers.base import DistributedDataClassifier
+from nemo_curator.datasets import DocumentDataset
+from nemo_curator.utils.aegis_utils import format_aegis
+
+
+@dataclass
+class AegisConfig:
+    peft_model_name_or_path: str
+    token: Optional[Union[str, bool]] = None
+    pretrained_model_name_or_path: str = "meta-llama/LlamaGuard-7b"
+    dtype: torch.dtype = torch.bfloat16
+    max_length: int = 4096
+
+
+AEGIS_LABELS = [
+    "unknown",
+    "safe",
+    "O1",
+    "O2",
+    "O3",
+    "O4",
+    "O5",
+    "O6",
+    "O7",
+    "O8",
+    "O9",
+    "O10",
+    "O11",
+    "O12",
+    "O13",
+]
+
+
+class AegisModel(nn.Module):
+    def __init__(
+        self,
+        pretrained_model_name_or_path: str,
+        peft_model_name_or_path: str,
+        dtype: torch.dtype,
+        token: str,
+    ):
+        super().__init__()
+        base_model = AutoModelForCausalLM.from_pretrained(
+            pretrained_model_name_or_path, torch_dtype=dtype, token=token
+        )
+        self.model = PeftModel.from_pretrained(base_model, peft_model_name_or_path)
+
+    @torch.no_grad()
+    def forward(self, batch):
+        response = self.model.generate(
+            **batch,
+            max_new_tokens=100,
+            pad_token_id=0,
+        )
+        return response
+
+
+class AegisHFModel(HFModel):
+    def __init__(self, config: AegisConfig):
+        self.config = config
+        super().__init__(
+            config.pretrained_model_name_or_path,
+            max_mem_gb=48,
+            start_batch_size=4,
+            end_batch_size=32,
+            batch_size_increment=4,
+            start_seq_len=1024,
+            seq_len_increment=1024,
+        )
+
+    def load_model(self, device="cuda"):
+        model = AegisModel(
+            self.config.pretrained_model_name_or_path,
+            self.config.peft_model_name_or_path,
+            self.config.dtype,
+            self.config.token,
+        )
+        model = model.to(device)
```

Comment:
Right now I'm getting an `OutOfMemoryError: CUDA out of memory. Tried to allocate 86.00 MiB. GPU` here. I've tried increasing `max_mem_gb` but still can't get it to run.

Reply:
I can confirm that:
```
export CUDA_VISIBLE_DEVICES="6,7" && aegis_classifier_inference \
    --input-data-dir /home/nfs/syurick/LLM_domain_classifier_inference/justext_resiliparse_trafilatura2 \
    --input-file-type jsonl \
    --output-data-dir /home/nfs/syurick/LLM_domain_classifier_inference/justext_resiliparse_trafilatura_aegis_output \
    --output-file-type jsonl \
    --device gpu
```
works for me (takes a long time to complete but I can see it iterating through each row of the JSONL file), so it looks like this is an issue with trying to run it in a Jupyter Notebook?

Reply:
Setting `%env CUDA_VISIBLE_DEVICES 6,7` in the notebook worked for me.